### PR TITLE
Added support for the Key Management API

### PR DESCRIPTION
**Pull request by rozza**
JAVA-4433

### CODE REVIEW

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assumptions.assumeFalse;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+// See: https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/tests#decryption-events
+public abstract class AbstractClientSideEncryptionDecryptionEventsTest {
+    private static final List<Bson> AGGREGATION_PIPELINE = singletonList(Aggregates.match(new BsonDocument()));
+    private MongoClient encryptedClient;
+    private ClientEncryption clientEncryption;
+    private Map<String, Map<String, Object>> kmsProviders;
```
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java:66]: IntelliJ suggests that this field does not need to be a field as it is used only in `setUp`. This is also applicable to `AbstractClientSideEncryptionUniqueIndexKeyAltNamesTest`.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    @AfterEach
+    public void cleanUp() {
+        if (clientEncryption != null) {
+            clientEncryption.close();
+        }
+        if (encryptedClient != null) {
+            encryptedClient.close();
+        }
```
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java:127]: When we close resources this way, we risk to not close some of them if closing a previous resource fails. I propose to use the same approach you used in `Crypt.close`. This is also applicable to `AbstractClientSideEncryptionUniqueIndexKeyAltNamesTest` and `AbstractClientSideEncryptionExplicitEncryptionTest`.

**Review by stIncMale — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 3]
```diff
@@ -73,6 +73,14 @@ configure(coreProjects) {
...
         google()
         mavenCentral()
+        // TODO temp remove once JAVA-4433
+        maven {
+            url 'https://oss.sonatype.org/content/repositories/snapshots'
+            content {
+                // Only include mongodb-crypt snapshots
+                includeGroup "org.mongodb"
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rozza) [reply build.gradle:83]: Done

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedClientEncryptionHelper.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+    private OperationResult resultOf(final Supplier<BsonValue> operationResult) {
+        try {
+            return OperationResult.of(operationResult.get());
+        } catch (Exception e) {
+            return OperationResult.of(e);
+        }
+    }
+
+    public static Map<String, Map<String, Object>> createKmsProvidersMap(final BsonDocument kmsProviders) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedClientEncryptionHelper.java:52]: They are similar in that they form the same map - however, the legacy tests are shaped differently. Hopefully, they will eventually be removed now that the unified tests have been written.

File: driver-sync/src/test/functional/com/mongodb/client/unified/ClientSideEncryptionTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            @SuppressWarnings("unused") final String testDescription,
+            final String schemaVersion, @Nullable final BsonArray runOnRequirements, final BsonArray entities,
+            final BsonArray initialData, final BsonDocument definition) {
+        super(schemaVersion, runOnRequirements, entities, initialData, definition);
+
+    }
+
+    @Parameterized.Parameters(name = "{0}: {1}")
+    public static Collection<Object[]> data() throws URISyntaxException, IOException {
+        return getTestData("client-side-encryption/unified");
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/ClientSideEncryptionTest.java:39]: Ack - will move

File: driver-core/src/main/com/mongodb/client/model/vault/RewrapManyDataKeyResult.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.mongodb.bulk.BulkWriteResult;
+import com.mongodb.lang.Nullable;
+
+/**
+ * The result of the rewrapping of data keys
+ *
+ * @since 4.7
+ */
+public class RewrapManyDataKeyResult {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/vault/RewrapManyDataKeyResult.java:27]: Thanks - made final

File: driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java [hunk 7]
```diff
@@ -67,6 +90,60 @@ public BsonValue decrypt(final BsonBinary value) {
...
+
+    @Override
+    public BsonDocument addKeyAltName(final BsonBinary id, final String keyAltName) {
+        return collection.findOneAndUpdate(Filters.eq("_id", id), Updates.addToSet("keyAltNames", keyAltName));
+    }
+
+    @Override
+    public BsonDocument removeKeyAltName(final BsonBinary id, final String keyAltName) {
+        return collection.findOneAndUpdate(Filters.eq("_id", id),
+                singletonList(BsonDocument.parse(format(UPDATE_TEMPLATE, keyAltName, keyAltName))));
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java:116]: Agreed -  that was me being lazy. Updated.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java (hunk 2 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java:127]: Done

File: driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java [hunk 8]
```diff
@@ -47,6 +47,10 @@ public void assertValuesMatch(final BsonValue expected, final BsonValue actual)
     private void assertValuesMatch(final BsonValue initialExpected, final @Nullable BsonValue actual,
             final @Nullable String keyContext, final int arrayPositionContext) {
         BsonValue expected = initialExpected;
+        if (expected.isNull() && actual == null) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java:52]: Removed EventMatcher.massageCommand

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java (hunk 2 above)
Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java]: Without `@SuppressWarnings("try")`, like in `Crypt.close`, the compiler is unhappy and spits out warnings.

File: driver-sync/src/test/functional/com/mongodb/client/unified/EventMatcher.java [hunk 9]
```diff
@@ -201,4 +201,5 @@ public static String getReasonString(final ConnectionClosedEvent.Reason reason)
                 throw new UnsupportedOperationException("Unsupported reason: " + reason);
         }
     }
+
```
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/unified/EventMatcher.java:204]: This new line (not at the end of file) is now the only change in the file. Let's roll it back.

File: driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java:52]: > Its kind of similar - the json for the test says the expectedResult is null but we parse it using BsonDocument.parse which makes it a BsonNull - so this check is manage the type conversion caused by our json parsing.

So `BsonDocument.parse("{\"f\": null}").get("f")` returns `BsonNull.VALUE`. This is expected and does not seem to be a peculiarity of parsing: the key exists and its value is `null`. `actual` is `null` and not `BsonNull.VALUE` only if the key does not exist. Based on https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#evaluating-matches (at least as far as I understand):
```
...
assert that actual is a document

for every key/value in expected:
  // handle key-based operators (e.g. $$exists, $$unsetOrMatches)
  if value is a document and its first and only key starts with "$$":
    execute any assertion(s) for the special operator
    continue to the next iteration unless actual value must be matched

    assert that actual[key] exists
    assert that actual[key] matches value
...
```
we must assert that the key exists and that its value is equal to the one specified. So it seems that if a specific unified test test is supposed to expect either BSON `null` values, or absent keys, it should express the expectation as `{$$unsetOrMatches: null}`. And if it specifies `null` as the expected value, then the key must be present. If we don't distinguish keys that exist with `null` value, and keys that don't exist in our unified test runner, then we seem to violate the unified test format spec.

File: driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java (hunk 1 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionDecryptionEventsTest.java]: Done.

File: driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java]: I find the document syntax overwhelmingly verbose and hard to parse. But I see your point the updates to the document are equally hard to parse. I think I'd like a parse method that can also accept bsonvalues for injection - similar to how slf4j handles logging templates.

Will update

File: driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java [hunk 10]
```diff
@@ -480,6 +496,41 @@ private void initBucket(final BsonDocument entity, final String id, final Functi
...
+                    break;
+                case "kmsProviders":
+                    builder.kmsProviders(createKmsProvidersMap(entry.getValue().asDocument()));
+                    break;
+                default:
+                    throw new UnsupportedOperationException("Unsupported client encryption option: " + entry.getKey());
+            }
+        }
+
+        ClientEncryption clientEncryption = new ClientEncryptionImpl(Assertions.notNull("mongoClient", mongoClient), builder.build());
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java:530]: This needs an indirection similar to what is done in `Entities#init`, so that when testing reactive streams we make use of `com.mongodb.reactivestreams.client.syncadapter.SyncClientEncryption`

**Review by jyemin — CHANGES_REQUESTED** [review]
Initial review of API and unified tests.

File: driver-sync/src/test/functional/com/mongodb/client/unified/ErrorMatcher.java [hunk 11]
```diff
@@ -56,10 +57,10 @@ void assertErrorsMatch(final BsonDocument expectedError, final Exception e) {
         }
         if (expectedError.containsKey("isClientError")) {
             assertEquals(context.getMessage("Exception must be of type MongoClientException or IllegalArgumentException"
-                            + " or IllegalStateException or MongoSocketException"),
+                            + " or IllegalStateException or MongoSocketException or MongoInternalException"),
                     expectedError.getBoolean("isClientError").getValue(),
                     e instanceof MongoClientException || e instanceof IllegalArgumentException || e instanceof IllegalStateException
-                            || e instanceof MongoSocketException);
+                            || e instanceof MongoSocketException || e instanceof MongoInternalException);
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/unified/ErrorMatcher.java:63]: Curious where the MongoInternalException is being thrown from, and why.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-sync/src/main/com/mongodb/client/internal/Crypt.java [hunk 12]
```diff
@@ -259,6 +284,8 @@ private RawBsonDocument executeStateMachine(final MongoCryptContext cryptContext
                     break;
                 case READY:
                     return cryptContext.finish();
+                case DONE:
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-sync/src/main/com/mongodb/client/internal/Crypt.java:287]: Kevin explained to me that this state is entered when no keys are matched

File: driver-core/src/main/com/mongodb/client/model/vault/DataKeyOptions.java [hunk 13]
```diff
@@ -117,11 +136,26 @@ public BsonDocument getMasterKey() {
...
+     * Gets the custom key material if set.
+     *
+     * <p>The optional BinData of 96 bytes to use as custom key material for the data key being created.
+     * If set the custom key material is used for encrypting and decrypting data. Otherwise, the key material for the new data key is
+     * generated from a cryptographically secure random device.</p>
+
+     * @return the custom key material for the data key or null
+     * @since 4.7
+     */
+    public byte[] getKeyMaterial() {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/vault/DataKeyOptions.java:149]: Marked all getters as nullable

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java [hunk 14]
```diff
@@ -199,6 +201,25 @@ public Mono<BsonValue> decryptExplicitly(final BsonBinary value) {
                 .map(result -> result.get("v"));
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java:204]: Removed since

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java [hunk 15]
```diff
@@ -221,9 +242,8 @@ private Mono<RawBsonDocument> executeStateMachine(final Supplier<MongoCryptConte
         try {
             MongoCryptContext cryptContext = cryptContextSupplier.get();
             return Mono.<RawBsonDocument>create(sink -> executeStateMachineWithSink(cryptContext, databaseName, sink))
-                    .doFinally(s -> cryptContext.close())
-                    .doOnError(t -> LOGGER.error(format("Crypt error: %s", t.getMessage())))
-                    .doOnError(MongoCryptException.class, this::wrapInClientException);
+                    .doOnError(this::wrapInClientException)
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java:245]: Because not all errors are MongoCryptException - for example FetchKeys throws a MongoException. And we now test that code path.

File: driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java [hunk 16]
```diff
@@ -480,6 +496,41 @@ private void initBucket(final BsonDocument entity, final String id, final Functi
         putEntity(id, gridFSBucketSupplier.apply(database), buckets);
     }
 
+    private void initClientEncryption(final BsonDocument entity, final String id) {
+        assumeTrue(hasEncryptionTestsEnabled());
+        assumeTrue(serverVersionAtLeast(4, 2));
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java:501]: I'll move to the csfle checker. See https://github.com/mongodb/specifications/commit/7536d8aa3589efa0d388e0d50b8e86908dc775c0

File: driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java [hunk 17]
```diff
@@ -480,6 +496,41 @@ private void initBucket(final BsonDocument entity, final String id, final Functi
         putEntity(id, gridFSBucketSupplier.apply(database), buckets);
     }
 
+    private void initClientEncryption(final BsonDocument entity, final String id) {
+        assumeTrue(hasEncryptionTestsEnabled());
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java:500]: Sounds good will move.

File: driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java (hunk 10 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java:530]: Great catch.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java [hunk 18]
```diff
@@ -228,7 +230,8 @@ private void assertOperation(final BsonDocument operation, final int operationIn
         context.push(ContextElement.ofCompletedOperation(operation, result, operationIndex));
         if (!operation.getBoolean("ignoreResultAndError", BsonBoolean.FALSE).getValue()) {
             if (operation.containsKey("expectResult")) {
-                assertNotNull(context.getMessage("The operation expects a result but an exception occurred"), result.getResult());
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java:231]: Checking a result being null is not the same as checking if there was an exception - so I corrected the assertion.

File: driver-sync/src/main/com/mongodb/client/vault/ClientEncryption.java [hunk 19]
```diff
@@ -78,6 +84,82 @@ public interface ClientEncryption extends Closeable {
...
+    BsonDocument getKeyByAltName(String keyAltName);
+
+    /**
+     * Decrypts multiple data keys and (re-)encrypts them with the current masterKey.
+     *
+     * @param filter the filter
+     * @return the result
+     * @since 4.7
+     */
+    RewrapManyDataKeyResult rewrapManyDataKey(BsonDocument filter);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/vault/ClientEncryption.java:151]: Done - although I noticed we don't have a `MongoClient#getCodecRegistry()` method but we do have it for the `MongoDatabase` and `MongoCollection`!

File: driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/Entities.java]: Done

File: driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java [hunk 20]
```diff
@@ -119,7 +119,6 @@ private void assertValuesMatch(final BsonValue initialExpected, final @Nullable
                                 throw new UnsupportedOperationException("Unsupported special operator: " + value.asDocument().getFirstKey());
                         }
                     }
-
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/ValueMatcher.java]: Reverted to Master's version

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypt.java]: Reverted - was due to not using the SyncClientEncryption

File: driver-sync/src/test/functional/com/mongodb/client/unified/ErrorMatcher.java (hunk 11 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/ErrorMatcher.java]: Reverted - was needed due to not using the SyncClientEncryption

**Review by jyemin — CHANGES_REQUESTED** [review]
Almost there.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java [hunk 21]
```diff
@@ -137,6 +139,8 @@ private static Object[] createTestData(final BsonDocument fileDocument, final Bs
 
     protected abstract GridFSBucket createGridFSBucket(MongoDatabase database);
 
+    protected abstract ClientEncryption createClientEncryption(MongoClient mongoClient, ClientEncryptionSettings clientEncryptionSettings);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java]: Done

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedSyncTest.java [hunk 22]
```diff
@@ -46,4 +49,9 @@ protected MongoClient createMongoClient(final MongoClientSettings settings) {
     protected GridFSBucket createGridFSBucket(final MongoDatabase database) {
         return GridFSBuckets.create(database);
     }
+
+    @Override
+    protected ClientEncryption createClientEncryption(final MongoClient mongoClient, final ClientEncryptionSettings clientEncryptionSettings) {
+        return new ClientEncryptionImpl(mongoClient, clientEncryptionSettings);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedSyncTest.java]: No - passing in a MongoClient is only available via the non public API.

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedReactiveStreamsTest.java [hunk 23]
```diff
@@ -50,4 +54,9 @@ protected MongoClient createMongoClient(final MongoClientSettings settings) {
     protected GridFSBucket createGridFSBucket(final MongoDatabase database) {
         return new SyncGridFSBucket(GridFSBuckets.create(((SyncMongoDatabase) database).getWrapped()));
     }
+
+    @Override
+    protected ClientEncryption createClientEncryption(final MongoClient mongoClient, final ClientEncryptionSettings clientEncryptionSettings) {
+        return new SyncClientEncryption(new ClientEncryptionImpl(((SyncMongoClient) mongoClient).getWrapped(), clientEncryptionSettings));
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedReactiveStreamsTest.java]: No - passing in a MongoClient is only available via the non public API

[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedReactiveStreamsTest.java]: It can be added to the public API but feel that would be a different ticket.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java:231]: Had to revert as some operations do not produce a result and then are checked against `{$unsetOrMatches: null}`

File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedReactiveStreamsTest.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/unified/UnifiedReactiveStreamsTest.java]: Yes, definitely a different ticket.  It was left out intentionally, so we would need to evaluate if it's a good idea to add it.  Taking just MongoClientSettings allows the ClientEncryptions to manage the lifetime of the key vault client (note the the close method closes the key vault client.  For the unified tests, that's ok because the lifetime of the the MongoClient and ClientEncryptions instances are the same, so we can get away with this.

It would probably work to grab the MongoClientSettings from the MongoClient (it will have the appropriate listeners), but it doesn't seem worth the effort.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java:231]: ah... ok, that was not a scenario I considered!
